C library: additional floating-point functions and cleanup

tautschnig opened the pull request:
Please see individual commit messages.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
C library: implement log2{,f,l} and log10{,f,l}

Follows the same approximation approach as previously taken for log (and
logf, logl).

tautschnig committed:
C library: fix logl in case long double == double

We previously ended up with recursion.

tautschnig committed:
C library: implement fma{,f,l}

Model fused multiply-add as documented in its man page.

tautschnig committed:
C library: implement feraiseexcept

Model floating-point exceptions as failing assertions.

tautschnig committed:
C library: do not flag division-by-zero when building NaN

We construct NaN (and Inf) by dividing by zero, which is a
standards-compliant way in that Nan (Inf) is the correct result for
these cases. Do not flag these operations as division-by-zero, which the
user would not expect.

tautschnig committed:
C library/fma: detect overflow

Fail as documented rather than via built-in assertions when overflowing.

tautschnig committed:
C library: do not report overflow in sqrt

We guess upper and lower bounds and check them for infinity afterwards.
Those multiplications should not be flagged by auto-generated
assertions.

kroening commented on src/ansi-c/library/math.c:
```diff
@@ -854,11 +899,14 @@ float sqrtf(float f)
+#pragma CPROVER check push
+#pragma CPROVER check disable "float-overflow"
     float lowerSquare = lower * lower;
     __CPROVER_assume(__CPROVER_isnormalf(lowerSquare));
```
A square root may well be denormal?

tautschnig replied:
I don't think that's possible for the case that `f` (which we are taking the square root of) is normal, which this branch assumes: given a denormal number is smaller than 1, any denormal square root `s` cannot be smaller than `f` (for $s < 1 \wedge s \geq 0 \implies s^2 \leq s$), so `f` has to be denormal when `s` is denormal.

tautschnig committed:
C library: fix _Static_assert for MSVC

Visual Studio knows static_assert, but not the standardised
_Static_assert.

tautschnig committed:
C library: do not flag float overflow with HUGE_VAL

Visual Studio uses `1.000000e+300 * 1.000000e+300` as value behind the
`HUGE_VAL` macro. We should not report an overflow for this approach of
producing infinity.
